flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java

```diff
@@ -0,0 +1,703 @@
+            FlinkBlueGreenDeployments deployments,
+            FlinkBlueGreenDeploymentStatus deploymentStatus,
+            DeploymentType currentDeploymentType,
+            Context<FlinkBlueGreenDeployment> josdkContext)
+            throws Exception {
+
+        if (hasSpecChanged(bgDeployment.getSpec(), deploymentStatus)) {
+
+            // Ack the change in the spec (setLastReconciledSpec)
+            setLastReconciledSpec(bgDeployment, deploymentStatus);
+
+            FlinkDeployment currentFlinkDeployment =
+                    DeploymentType.BLUE == currentDeploymentType
+                            ? deployments.getFlinkDeploymentBlue()
+                            : deployments.getFlinkDeploymentGreen();
+
+            if (isDeploymentReady(currentFlinkDeployment, josdkContext, deploymentStatus)) {
+
+                DeploymentType nextDeploymentType = DeploymentType.BLUE;
+                FlinkBlueGreenDeploymentState nextState =
+                        FlinkBlueGreenDeploymentState.TRANSITIONING_TO_BLUE;
+                FlinkResourceContext<FlinkDeployment> resourceContext =
+                        ctxFactory.getResourceContext(currentFlinkDeployment, josdkContext);
+
+                // Updating status
+                if (DeploymentType.BLUE == currentDeploymentType) {
+                    nextState = FlinkBlueGreenDeploymentState.TRANSITIONING_TO_GREEN;
+                    nextDeploymentType = DeploymentType.GREEN;
+                }
+
+                Savepoint lastCheckpoint = configureSavepoint(resourceContext);
+
+                return initiateDeployment(
+                        bgDeployment,
+                        deploymentStatus,
+                        nextDeploymentType,
+                        nextState,
+                        lastCheckpoint,
+                        josdkContext,
+                        false);
+            } else {
+                // If the current running FlinkDeployment is not in RUNNING/STABLE,
+                // we flag this Blue/Green as FAILING
+                if (deploymentStatus.getJobStatus().getState() != JobStatus.FAILING) {
+                    return patchStatusUpdateControl(
+                            bgDeployment, deploymentStatus, null, JobStatus.FAILING);
+                }
+            }
+        }
+
+        return UpdateControl.noUpdate();
+    }
+
+    private static void setLastReconciledSpec(
+            FlinkBlueGreenDeployment bgDeployment,
+            FlinkBlueGreenDeploymentStatus deploymentStatus) {
+        deploymentStatus.setLastReconciledSpec(
+                SpecUtils.serializeObject(bgDeployment.getSpec(), "spec"));
+        deploymentStatus.setLastReconciledTimestamp(System.currentTimeMillis());
+    }
+
+    public void logPotentialWarnings(
+            FlinkDeployment flinkDeployment,
+            Context<FlinkBlueGreenDeployment> josdkContext,
+            long lastReconciliationTimestamp) {
```

Comment:
We are already logging these errors as part of the FlinkDeployment logic, this is duplicated code/logic here that should be removed